lib/piezo.js

```diff
@@ -129,202 +137,284 @@ Piezo.Notes = {
- * supported and won't crash the board.
- * @return bool
- */
-function isValidOctave(octave) {
-  // octave should be a number.
-   if (octave && !isNaN(octave) &&
-     (octave >= 0 && octave <= 8)) {
-       return true;
-   }
-
-   return false;
+  return octave;
 }
 
-Piezo.prototype.defaultOctave = function(octave) {
-  if (typeof this.defaultOctave.value === "undefined") {
-    this.defaultOctave.value = 4;
+function normalizeNote(note, defaults) {
+  if (!note) {
+    return null;
   }
 
-  if (isValidOctave(octave)) {
-    this.defaultOctave.value = octave;
+  if (typeof note !== "string") {
+    return note;
   }
 
-  return this.defaultOctave.value;
-};
+  defaults = defaults || {
+    octave: "4",
+    beat: 1/2
+  };
 
-Piezo.prototype.note = function(note, duration) {
-  var tone = parseTone(note);
+  var normalizedNote = {};
 
-  return this.tone(tone, duration);
-};
+  if (note.length === 1) {
+    normalizedNote.notation = note[0];
+    normalizedNote.beat = defaults.beat;
+    normalizedNote.octave = defaults.octave;
 
-Piezo.prototype.tone = function(tone, duration) {
-  if (isNaN(tone) || isNaN(duration)) {
-    // Very Bad Things happen if one tries to play a NaN tone
-    throw new Error(
-      "Piezo.tone: invalid tone or duration"
-    );
+    return normalizedNote;
   }
 
-  clearTimer.call(this);
+  if (note.length === 3) {
+    normalizedNote.beat = normalizeBeat(note[0]) || defaults.beat;
+    normalizedNote.notation = note[1];
+    normalizedNote.octave = normalizeOctave(note[2]) || defaults.octave;
 
-  var timer = this.timer = new Timer();
-  var value = 1;
+    return normalizedNote;
+  }
 
-  timer.setInterval(function() {
-    value = value === 1 ? 0 : 1;
-    this.io.digitalWrite(this.pin, value);
+  if (note.length === 2) {
+    if (isNaN(note[0])) {
+      normalizedNote.notation = note[0];
+      normalizedNote.octave = note[1];
+      normalizedNote.beat = defaults.beat;
+    } else {
+      normalizedNote.beat = normalizeBeat(note[0]);
+      normalizedNote.octave = defaults.octave;
+      normalizedNote.notation = note[1];
+    }
 
-    if ((timer.difTime / 1000000) > duration) {
-      clearTimer.call(this);
+    return normalizedNote;
+  }
+}
+
+function normalizeMelody(melody, defaults) {
+  if (!melody) {
+    return null;
+  }
+
+  if (typeof melody === "string") {
+    var melodyObject = {
+      sheet: melody
+    };
+
+    melody = melodyObject;
+  }
+
+  if (typeof melody.sheet !== "string") {
+    return null;
+  }
+
+  defaults = defaults || {
+    octave: 4,
+    duration: 500,
+    beat: 1 / 2,
+    tempo: 125
+  };
+
+  // 1. remove colons and semi-colons
+  // 2. remove additional whitespace
+  // 3. clean up defaults if present.
+  melody.sheet = melody.sheet.replace(/,|:/g, " ");
+  melody.sheet = melody.sheet.replace(/\s{2,}/g, " ");
+  melody.sheet = melody.sheet.replace(/\s{1,}=|=\s{1,}/g, "=");
+
+  // Ensure we have defaults properly set up.
+  melody.notes = melody.notes = [];
```

Comment:
`melody.notes = melody.notes = [];`? Should this be: `melody.notes = melody.notes || [];`?